Generalize host allocator to be device-agnostic

guangyey opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #123080
* __->__ #123079

# Motivation
According to [[RFC] Intel GPU Runtime Upstreaming for Allocator](https://github.com/pytorch/pytorch/issues/116322), we would like to generalize device and host allocator to be device-agnostic. We prioritize the host allocator as it is simpler and more native than the device allocator. In this PR, we intend to refactor the host allocator to make it be shared across different backends. In 2nd PR, we will support host allocator on XPU backend. 

# Design
The previous design:
- `CUDAHostAllocatorWrapper` inherits from `c10::Allocator`, and `CUDAHostAllocator` is an implementation of `CUDAHostAllocatorWrapper`.

The design in this PR:
- `CachingHostAllocatorImpl` is an interface that implements the caching host allocator logic that can be sharable across each backend.
- `CachingHostAllocatorInterface` inherits from `c10::Allocator` as an interface and accepts `CachingHostAllocatorImpl` as its implementation.
- `CUDACachingHostAllocator` is a CUDA host allocator whose implementation is `CUDACachingHostAllocatorImpl` which is specialized from `CachingHostAllocatorImpl`.

This design can 
- share most code of caching mechanism across different backends, and
- keep the flexibility to expand its exclusive feature on each backend.

# Additional Context
In addition, we will continue to generalize the device allocator in the next stage.



cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

guangyey committed:
generalize host allocator

[ghstack-poisoned]

guangyey committed:
Update on "[WIP] Generalize host allocator to be device-agnostic"




[ghstack-poisoned]

guangyey committed:
Update on "[WIP] Generalize host allocator to be device-agnostic"




[ghstack-poisoned]

guangyey committed:
Update on "[WIP] Generalize host allocator to be device-agnostic"




[ghstack-poisoned]

guangyey committed:
Update on "[WIP] Generalize host allocator to be device-agnostic"




[ghstack-poisoned]

guangyey committed:
Update on "[WIP] Generalize host allocator to be device-agnostic"




[ghstack-poisoned]

guangyey committed:
Update on "[WIP] Generalize host allocator to be device-agnostic"




[ghstack-poisoned]

guangyey committed:
Update on "[WIP] Generalize host allocator to be device-agnostic"




[ghstack-poisoned]

guangyey committed:
Update on "[WIP] Generalize host allocator to be device-agnostic"




[ghstack-poisoned]

guangyey committed:
Update on "[WIP] Generalize host allocator to be device-agnostic"




[ghstack-poisoned]

guangyey committed:
Update on "[WIP] Generalize host allocator to be device-agnostic"




[ghstack-poisoned]

jgong5 commented on aten/src/ATen/core/CachingHostAllocator.h:
```diff
@@ -0,0 +1,354 @@
+ * list. Remove all element of free list, then remove them from block list and
+ * release the associated pinned memory allocation via free_block.
+ *
+ * To share the mechanism across each backend, we abstract its logic to be
+ * device-agnostic through an interface. Then each backend can specialize
+ * their own implement and the related runtime functions, like
+ * allocate_host_memory, free_block, record_stream, query_event; Even inheriting
+ * from the interface to extend additional features.
```
Can we make it more clear what a backend implementation needs to do, e.g., creating a class `xyz` that derives `CachingHostAllocatorImplInterface`, specifying template args (can you add explanation to the temp args too?), implementing methods `abc`... ,etc.?

guangyey replied:
refine comments and add more notes.

jgong5 reviewed (approved):
Overall LGTM. Suggest to add more notes on what a backend needs to do to add its own host allocator.

EikanWang commented on aten/src/ATen/core/CachingHostAllocator.h:
```diff
@@ -0,0 +1,381 @@
+    typename E,
+    typename B = HostBlock<S>,
+    typename C = ComparatorSize<B>>
+struct CachingHostAllocatorImplInterface {
```
`Caching` is the detailed implementation. Should we just rename the class name as `HostAllocatorImplInterface`?

guangyey committed:
Update on "[WIP] Generalize host allocator to be device-agnostic"


# Motivation
According to [[RFC] Intel GPU Runtime Upstreaming for Allocator](https://github.com/pytorch/pytorch/issues/116322), we would like to generalize device and host allocator to be device-agnostic. We prioritize the host allocator as it is simpler and more native than the device allocator. In this PR, we intend to refactor the host allocator to make it be shared across different backends. In 2nd PR, we will support host allocator on XPU backend. 

# Design
The previous design:
- `CUDAHostAllocatorWrapper` inherits from `c10::Allocator`, and `CUDAHostAllocator` is an implementation of `CUDAHostAllocatorWrapper`.

The design in this PR:
- `CachingHostAllocatorImplInterface` is an interface that implements the caching host allocator logic that can be sharable across each backend.
- `HostAllocatorInterface` inherits from `c10::Allocator` as an interface and accepts `CachingHostAllocatorImplInterface` as its implementation.
- `CUDAHostAllocator` is a CUDA host allocator whose implementation is `CUDAHostAllocatorImpl` which is specialized from `CachingHostAllocatorImplInterface`.

This design can 
- share most code of caching mechanism across different backends, and
- keep the flexibility to expand its exclusive feature on each backend.

# Additional Context
In addition, we will continue to generalize the device allocator in the next stage.



cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

guangyey committed:
Update on "[WIP] Generalize host allocator to be device-agnostic"


# Motivation
According to [[RFC] Intel GPU Runtime Upstreaming for Allocator](https://github.com/pytorch/pytorch/issues/116322), we would like to generalize device and host allocator to be device-agnostic. We prioritize the host allocator as it is simpler and more native than the device allocator. In this PR, we intend to refactor the host allocator to make it be shared across different backends. In 2nd PR, we will support host allocator on XPU backend. 

# Design
The previous design:
- `CUDAHostAllocatorWrapper` inherits from `c10::Allocator`, and `CUDAHostAllocator` is an implementation of `CUDAHostAllocatorWrapper`.

The design in this PR:
- `CachingHostAllocatorImplInterface` is an interface that implements the caching host allocator logic that can be sharable across each backend.
- `HostAllocatorInterface` inherits from `c10::Allocator` as an interface and accepts `CachingHostAllocatorImplInterface` as its implementation.
- `CUDAHostAllocator` is a CUDA host allocator whose implementation is `CUDAHostAllocatorImpl` which is specialized from `CachingHostAllocatorImplInterface`.

This design can 
- share most code of caching mechanism across different backends, and
- keep the flexibility to expand its exclusive feature on each backend.

# Additional Context
In addition, we will continue to generalize the device allocator in the next stage.



cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

guangyey committed:
Update on "[WIP] Generalize host allocator to be device-agnostic"


# Motivation
According to [[RFC] Intel GPU Runtime Upstreaming for Allocator](https://github.com/pytorch/pytorch/issues/116322), we would like to generalize device and host allocator to be device-agnostic. We prioritize the host allocator as it is simpler and more native than the device allocator. In this PR, we intend to refactor the host allocator to make it be shared across different backends. In 2nd PR, we will support host allocator on XPU backend. 

# Design
The previous design:
- `CUDAHostAllocatorWrapper` inherits from `c10::Allocator`, and `CUDAHostAllocator` is an implementation of `CUDAHostAllocatorWrapper`.

The design in this PR:
- `CachingHostAllocatorImplInterface` is an interface that implements the caching host allocator logic that can be sharable across each backend.
- `HostAllocatorInterface` inherits from `c10::Allocator` as an interface and accepts `CachingHostAllocatorImplInterface` as its implementation.
- `CUDAHostAllocator` is a CUDA host allocator whose implementation is `CUDAHostAllocatorImpl` which is specialized from `CachingHostAllocatorImplInterface`.

This design can 
- share most code of caching mechanism across different backends, and
- keep the flexibility to expand its exclusive feature on each backend.

# Additional Context
In addition, we will continue to generalize the device allocator in the next stage.



cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

albanD reviewed (approved):
Is this actually used by default on CUDA today? Do you plan on using it by default for XPU?

Also reviewed assuming all the code was just moved. Did you actually do any logic change beyond adding the appropriate API? If not, SGTM!

guangyey committed:
Update on "[WIP] Generalize host allocator to be device-agnostic"


# Motivation
According to [[RFC] Intel GPU Runtime Upstreaming for Allocator](https://github.com/pytorch/pytorch/issues/116322), we would like to generalize device and host allocator to be device-agnostic. We prioritize the host allocator as it is simpler and more native than the device allocator. In this PR, we intend to refactor the host allocator to make it be shared across different backends. In 2nd PR, we will support host allocator on XPU backend. 

# Design
The previous design:
- `CUDAHostAllocatorWrapper` inherits from `c10::Allocator`, and `CUDAHostAllocator` is an implementation of `CUDAHostAllocatorWrapper`.

The design in this PR:
- `CachingHostAllocatorImplInterface` is an interface that implements the caching host allocator logic that can be sharable across each backend.
- `HostAllocatorInterface` inherits from `c10::Allocator` as an interface and accepts `CachingHostAllocatorImplInterface` as its implementation.
- `CUDAHostAllocator` is a CUDA host allocator whose implementation is `CUDAHostAllocatorImpl` which is specialized from `CachingHostAllocatorImplInterface`.

This design can 
- share most code of caching mechanism across different backends, and
- keep the flexibility to expand its exclusive feature on each backend.

# Additional Context
In addition, we will continue to generalize the device allocator in the next stage.



cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

guangyey committed:
Update on "Generalize host allocator to be device-agnostic"


# Motivation
According to [[RFC] Intel GPU Runtime Upstreaming for Allocator](https://github.com/pytorch/pytorch/issues/116322), we would like to generalize device and host allocator to be device-agnostic. We prioritize the host allocator as it is simpler and more native than the device allocator. In this PR, we intend to refactor the host allocator to make it be shared across different backends. In 2nd PR, we will support host allocator on XPU backend. 

# Design
The previous design:
- `CUDAHostAllocatorWrapper` inherits from `c10::Allocator`, and `CUDAHostAllocator` is an implementation of `CUDAHostAllocatorWrapper`.

The design in this PR:
- `CachingHostAllocatorImplInterface` is an interface that implements the caching host allocator logic that can be sharable across each backend.
- `HostAllocatorInterface` inherits from `c10::Allocator` as an interface and accepts `CachingHostAllocatorImplInterface` as its implementation.
- `CUDAHostAllocator` is a CUDA host allocator whose implementation is `CUDAHostAllocatorImpl` which is specialized from `CachingHostAllocatorImplInterface`.

This design can 
- share most code of caching mechanism across different backends, and
- keep the flexibility to expand its exclusive feature on each backend.

# Additional Context
In addition, we will continue to generalize the device allocator in the next stage.



cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

guangyey commented:
> Is this actually used by default on CUDA today? Do you plan on using it by default for XPU?
> 
> Also reviewed assuming all the code was just moved. Did you actually do any logic change beyond adding the appropriate API? If not, SGTM!

Yes, this caching mechanism is used on CUDA by default, and XPU will follow it.
I do **NOT** change any logic, just add some appropriate APIs to decouple its implementation with CUDA. To make another backend, like XPU, can share these pieces of code.

guangyey commented:
@pytorchbot merge
